File: services/tracking/src/shared/metrics/cloudwatch_metrics.py
"""CloudWatch custom-metrics publisher.

The wiring mirrors `shared/messaging/sqs_event_publisher.py` exactly: a Protocol
port, a concrete boto3-backed implementation, a no-op binding for suites that
must not reach AWS, and an `lru_cache`d factory keyed on PRIMITIVES (never on
`Settings`, which is unhashable — see `_cached_publisher` below). `shared/di/` in
this service is an empty placeholder; there is no framework container to register
into.

## One responsibility

This module turns a `(name, value, dimensions)` triple into one `PutMetricData`
call and nothing else. The SCHEDULING and the QUERIES live in
`features/tracking/commands/publish_metrics.py`. That split is what makes this
class unit-testable with a recording double and the loop testable with an
injected fake sleep.

## Failure policy: LOG AND SWALLOW, deliberately

`publish` never raises. A metrics backend being unreachable may never break the
request or the loop that produced the metric — the same stance
`SqsEventPublisher` takes for events, and for the same reason: the metric is a
secondary observation of work that already happened. This is NOT silent: every
failure is an `error` line carrying `app_event=metric_publish_failed`, which is
what makes it alertable.

## The namespace and the dimensions are a contract

Every 3MRAI metric, in every service, is published under the single namespace
`3MRAI`. The dimension SET is equally load-bearing: Floci does not aggregate
across dimensions, so the collector's `GetMetricData` query must name the exact
same set the datum was published with — a query that omits one returns
`Values: []` with `StatusCode: "Complete"`, a silent empty result rather than an
error. Dimensions are therefore low-cardinality labels only; never a user id, an
email or an order id.
"""

from __future__ import annotations

import logging
from functools import lru_cache
from typing import Any, Protocol

import boto3

from src.shared.config.settings import get_settings

logger = logging.getLogger(__name__)

#: The one namespace every 3MRAI metric is published under, across all four
#: services. Never a per-service namespace.
METRICS_NAMESPACE = "3MRAI"

#: The `Service` dimension value every metric from THIS service carries.
SERVICE_DIMENSION = "tracking"


class MetricsPublisher(Protocol):
    """Port for publishing one metric datum.

    `publish` NEVER raises — that is part of the contract, not an implementation
    detail, and every caller relies on it.
    """

    def publish(self, name: str, value: float, dimensions: dict[str, str]) -> None: ...


class CloudWatchMetricsPublisher:
    """Publishes to CloudWatch (Floci locally) via boto3.

    **Blocking.** boto3 is synchronous, so an async caller must invoke `publish`
    through `asyncio.to_thread` rather than inline — doing it on the event loop
    would stall the health check and every REST handler for the duration of the
    HTTP round trip.
    """

    def __init__(self, *, client: Any) -> None:
        self._client = client

    def publish(self, name: str, value: float, dimensions: dict[str, str]) -> None:
        """Emit one datum. Never raises — see the module docstring's policy.

        `value` is published as given, `0` included: a series that stops being
        published reads as "no data" in a dashboard, not as zero, so there is no
        falsy short-circuit here.
        """
        try:
            self._client.put_metric_data(
                Namespace=METRICS_NAMESPACE,
                MetricData=[
                    {
                        "MetricName": name,
                        "Value": value,
                        "Unit": "Count",
                        # CloudWatch's list-of-{Name,Value}, in the mapping's
                        # own (insertion) order. The collector's query must name
                        # this exact set — see the module docstring.
                        "Dimensions": [
                            {"Name": key, "Value": val}
                            for key, val in dimensions.items()
                        ],
                    }
                ],
            )
        except Exception:  # noqa: BLE001 - swallowed on purpose, see the docstring
            logger.exception(
                "metric_publish_failed",
                extra={
                    "app_event": "metric_publish_failed",
                    "reason": "put_metric_data_failed",
                    "metric_name": name,
                },
            )


class NoopMetricsPublisher:
    """Binding for suites (and runtimes) that must not reach CloudWatch."""

    def publish(self, name: str, value: float, dimensions: dict[str, str]) -> None:
        return None


@lru_cache(maxsize=1)
def _cached_publisher(
    endpoint_url: str | None, region: str
) -> CloudWatchMetricsPublisher:
    """One publisher (hence one boto3 client) per process, keyed on PRIMITIVES.

    Keyed on the values rather than on `Settings` for the reason recorded in
    `shared/db/engine.py`, `shared/grpc/users_client.py` and
    `shared/messaging/sqs_event_publisher.py`: pydantic's `BaseSettings` is
    unhashable, so an `lru_cache` taking a settings object raises `TypeError` on
    its very first call — and that failure is invisible to a test suite that
    injects its own double everywhere.
    """
    client = boto3.client("cloudwatch", endpoint_url=endpoint_url, region_name=region)
    return CloudWatchMetricsPublisher(client=client)


def shared_metrics_publisher() -> CloudWatchMetricsPublisher:
    """The process-wide publisher, built lazily from settings.

    Lazy, not module-level, so importing this module neither constructs a boto3
    client nor requires a valid environment — the same rule
    `shared_event_publisher()` and `shared_users_client()` follow, and what lets
    the test suite import the metrics loop without an env file.
    """
    settings = get_settings()
    return _cached_publisher(settings.aws_endpoint_url, settings.aws_region)

File: services/users/src/shared/metrics/cloudwatch-metrics.ts
import { CloudWatchClient, PutMetricDataCommand } from "@aws-sdk/client-cloudwatch";
import { SpanKind, SpanStatusCode, trace } from "@opentelemetry/api";
import { appLogger } from "../logging/app-logger.ts";

/** The one namespace every 3MRAI metric is published under. */
export const METRICS_NAMESPACE = "3MRAI";

// Own tracer, named for its area — same convention as publish-tracing's
// `users-messaging`.
const tracer = trace.getTracer("users-metrics");

/**
 * Publishes custom metrics to CloudWatch.
 *
 * Every failure is logged and swallowed: an unreachable metrics backend must never
 * fail the registration, login or password reset that triggered the metric.
 */
export class MetricsPublisher {
  private readonly client: CloudWatchClient;

  constructor({ client }: { client: CloudWatchClient }) {
    this.client = client;
  }

  async publish(
    name: string,
    value: number,
    dimensions: Record<string, string>,
    unit: "Count" | "Milliseconds" = "Count",
  ): Promise<void> {
    // CONTRACT: Name the span after the metric. The AWS SDK's auto span is called
    // `CloudWatch.PutMetricData` with no metric name, producing thousands of identical
    // unreadable bars. Do NOT suppress that child span — that means configuring the
    // SDK's instrumentation in code, which has silently broken telemetry three times
    // here. `publish` swallows every failure, so the span records the call's outcome
    // while the method returns normally. See [[logging-context]]
    await tracer.startActiveSpan(
      `cloudwatch PutMetricData ${name}`,
      {
        kind: SpanKind.CLIENT,
        attributes: {
          "rpc.system": "aws-api",
          "rpc.service": "CloudWatch",
          "rpc.method": "PutMetricData",
          "metric.name": name,
        },
      },
      async (span) => {
        try {
          await this.client.send(
            new PutMetricDataCommand({
              Namespace: METRICS_NAMESPACE,
              MetricData: [
                {
                  MetricName: name,
                  Value: value,
                  Unit: unit,
                  // CloudWatch's list-of-{Name,Value} shape. The exact dimension SET
                  // is load-bearing: a query naming a different set comes back EMPTY
                  // with StatusCode "Complete" rather than as an error.
                  Dimensions: Object.entries(dimensions).map(([Name, Value]) => ({ Name, Value })),
                },
              ],
            }),
          );
          span.setStatus({ code: SpanStatusCode.OK });
        } catch (err) {
          // Swallowed on purpose — see the class docstring.
          span.setStatus({
            code: SpanStatusCode.ERROR,
            message: err instanceof Error ? err.message : String(err),
          });
          appLogger.warn(
            {
              app_event: "metric_publish_failed",
              reason: err instanceof Error ? err.message : String(err),
              metric_name: name,
            },
            "failed to publish metric",
          );
        } finally {
          span.end();
        }
      },
    );
  }
}
